[SPARK-38664][CORE] Support compact EventLog when there are illegal characters in the path

lw33 opened the pull request:
### What changes were proposed in this pull request?
Do not escape illegal characters.

### Why are the changes needed?
Compact EventLog fail.

### Does this PR introduce any user-facing change?
No.

### How was this patch tested?
Unit Test

AmplabJenkins commented:
Can one of the admins verify this patch?

HyukjinKwon commented:
cc @HeartSaVioR FYI

LuciferYang commented on core/src/test/scala/org/apache/spark/deploy/history/EventLogFileCompactorSuite.scala:
```diff
@@ -119,7 +119,7 @@ class EventLogFileCompactorSuite extends SparkFunSuite {
       val fs = new Path(dir.getAbsolutePath).getFileSystem(hadoopConf)
 
-      val fileStatuses = writeEventsToRollingWriter(fs, "app", dir, sparkConf, hadoopConf,
+      val fileStatuses = writeEventsToRollingWriter(fs, "app a", dir, sparkConf, hadoopConf,
```
I check this case as follows:

```
gh pr checkout 35979
manually revert change of EventLogFileCompactor.scala
mvn clean install -DskipTests -pl core -am
mvn test -pl core -Dtest=none -DwildcardSuites=org.apache.spark.deploy.history.EventLogFileCompactorSuite
```
Test result as follows:
```
Discovery starting.
Discovery completed in 1 second, 647 milliseconds.
Run starting. Expected test count is: 9
EventLogFileCompactorSuite:
- No event log files
- No compact file, less origin files available than max files to retain
- No compact file, more origin files available than max files to retain
- compact file exists, less origin files available than max files to retain
- compact file exists, number of origin files are same as max files to retain
- compact file exists, more origin files available than max files to retain
- events for finished job are dropped in new compact file
- Don't compact file if score is lower than threshold
- rewrite files with test filters
Run completed in 2 seconds, 481 milliseconds.
Total number of tests run: 9
Suites: completed 2, aborted 0
Tests: succeeded 9, failed 0, canceled 0, ignored 0, pending 0
All tests passed.
```

It seems that the new UT is successful without this pr. @lw33 

Can you help double check? @HyukjinKwon

lw33 replied:
My mistake，i fixed this. Plz check it again.

LuciferYang replied:
On second thought, I think in the production environment, illegal characters(such as %) may appear in `basedir`, maybe we can add a new case as follows:

```scala
test("SPARK-38664: ${Test display name}") {
    // create a base dir with illegal characters
    val dir = Utils.createTempDir("spark-%")
    try {
      // do test 
    } finally {
      Utils.deleteRecursively(dir)
    }
  }

```

lw33 committed:
[SPARK-38664][Core] Support compact EventLog when there are illegal characters in the path

### What changes were proposed in this pull request?
Do not escape illegal characters.

### Why are the changes needed?
Compact EventLog fail.

### Does this PR introduce any user-facing change?
No.

### How was this patch tested?
Unit Test

LuciferYang commented on core/src/test/scala/org/apache/spark/deploy/history/EventLogFileCompactorSuite.scala:
```diff
@@ -284,6 +286,20 @@ class EventLogFileCompactorSuite extends SparkFunSuite {
   }
 
+  test("SPARK-38664: Support compact EventLog when there are illegal characters in the path") {
+    withTempDir { dir =>
```
If you still want to use `withTempDir`, you need to refactor `withTempDir` to accept `namePrefix`(but this will involve many file changes). 

At present, two `tmpdirs` are created in this way... which is not necessary

LuciferYang replied:
Ignore previous comments

LuciferYang commented:
LGTM + 1

lw33 commented:
@HeartSaVioR cc

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/deploy/history/EventLogFileCompactor.scala:
```diff
@@ -221,5 +221,5 @@ private class CompactedEventLogFileWriter(
     hadoopConf: Configuration)
   extends SingleEventLogFileWriter(appId, appAttemptId, logBaseDir, sparkConf, hadoopConf) {
 
-  override val logPath: String = originalFilePath.toUri.toString + EventLogFileWriter.COMPACTED
```
Is this safe with S3 filesystem?

viirya replied:
This looks like to have unexpected regression, if we are not sure the change is compatible with other file systems.

dongjoon-hyun replied:
Ya, it looks like that to me. Especially with Apache Hadoop 3.3.2.

HyukjinKwon replied:
Is it because the string presentation of the path is able to omit the scheme of the URI, or it does not encode the special characters to the encoded chars? Actually, we would have to always use Hadoop's `Path` to work on Hadoop's `Path` so `fs.default.name` is respected.

The only exception would be when this path is stored somewhere outside. In this case, the URI has to be stored as the fully-qualified one so different `fs.default.name` does not affect the original path.

This path URI stuff has many weird holes so we're just delegating to Hadoop's `Path` implementation, and leveraging their bug fixes.

I am fine if this change does not handle all of the cases but wanted to point out this direction is better than manually handling it in Spark.

viirya reviewed (commented):
Why this is `[SS]`? Is this for structured streaming?

dongjoon-hyun reviewed (changes requested):
Sorry guys. Supporting `illegal char` by removing `toURI` doesn't look like a safe improvement to me.

Given the trade-off between benefit and risk, we had better recommend to avoid this kind of **illegal char** usage instead, @lw33 .

lw33 commented:
> Sorry guys. Supporting `illegal char` by removing `toURI` doesn't look like a safe improvement to me.
> 
> Given the trade-off between benefit and risk, we had better recommend to avoid this kind of **illegal char** usage instead, @lw33 .

Why we need `toURI` here, i see the parent class `SingleEventLogFileWriter` `logpath` is `Path.toString`.

dongjoon-hyun commented:
Back to the original proposal, why do we need to support `illegal char`, @lw33 ? It's illegal, isn't it?

lw33 commented:
Yes, maybe we don't need to do this change. I just found this problem when compacting event log, the event log could write to the path, but compat failed, so I thought this might be a bug. Of course I could change the path to avoid this problem. @dongjoon-hyun

dongjoon-hyun commented:
Yes, I believe `illegal char` is the root cause of this problem, not Apache Spark, @lw33 . 
Apache Hadoop provides an abstraction layer for many file systems including local file system, S3A, WEBHDFS, FTP, and so on. In addition, there are many additional custom file systems which is unknown to us. The main concern about `illegal chars` is that some file systems may be broke due to that. Hence, I believe official support of illegal chars is not desirable for our end goal. That could open a can of worms in the end. As a perspective of committer (a maintainer of Apache Spark repo), I want to avoid that kind of situation.

If you don't mind, could you close this PR and issue, please?

dongjoon-hyun commented:
Thank you for your closing, @lw33 .
